File: packages/openapi/src/contracts/notification.ts
import { initContract } from "@ts-rest/core";
import { z } from "zod";
import { ZGetNotificationsQuery, ZGetNotificationsResponse } from "@nexus/zod";
import { getSecurityMetadata } from "@/utils.js";

const c = initContract();

export const notificationContract = c.router(
  {
    list: {
      summary: "List the authenticated user's notifications",
      method: "GET",
      path: "/notifications",
      query: ZGetNotificationsQuery,
      responses: {
        200: ZGetNotificationsResponse,
      },
      metadata: getSecurityMetadata(),
    },
    markRead: {
      summary: "Mark a notification as read",
      method: "POST",
      path: "/notifications/:id/read",
      pathParams: z.object({ id: z.string().uuid() }),
      body: c.noBody(),
      responses: {
        204: c.noBody(),
      },
      metadata: getSecurityMetadata(),
    },
    markAllRead: {
      summary: "Mark all notifications as read",
      method: "POST",
      path: "/notifications/read-all",
      body: c.noBody(),
      responses: {
        204: c.noBody(),
      },
      metadata: getSecurityMetadata(),
    },
  },
  { pathPrefix: "/api/v1" }
);

File: apps/backend/internal/router/notification.go
package router

import (
	"github.com/labstack/echo/v4"
	"github.com/sarbojitrana/nexus/internal/handler"
	"github.com/sarbojitrana/nexus/internal/middleware"
)

func registerNotificationRoutes(g *echo.Group, h *handler.Handlers, mw *middleware.Middlewares) {
	g.GET("/notifications", h.Notification.List, mw.Auth.RequireAuth)
	g.POST("/notifications/:id/read", h.Notification.MarkRead, mw.Auth.RequireAuth)
	g.POST("/notifications/read-all", h.Notification.MarkAllRead, mw.Auth.RequireAuth)
}
